### ISSUE TITLE
(1, n_samples) is not accepted as a mono waveform

### ISSUE THREAD

**Issue by dpwe** [librosa/librosa#1353]
**Description**
Librosa typically operates on one-dimensional, one-channel signals whose shape is `(n_samples,)`.  In some cases, stereo signals are accepted in "time-major" format, e.g. of shape `(2, n_samples)`.  But the analogous single-channel signal, whose shape is `(1, n_samples)`, is rejected -- at least by `numpy.resample()`

**To Reproduce**

```
import librosa
import numpy as np

# Conventional mono
mono_signal = np.random.randn(1000)
resampled_mono = librosa.resample(mono_signal, 1000, 2000)

# Conventional stereo
stereo_signal = np.random.randn(2, 1000)
resampled_stereo = librosa.resample(stereo_signal, 1000, 2000)

# Stereo-like single channel
onechan_signal = np.random.randn(1, 1000)
resampled_onechan = librosa.resample(onechan_signal, 1000, 2000)

```

**Expected behavior**
`resampled_mono == resampled_onechan[0]`

**Actual behavior**
```
---------------------------------------------------------------------------
ParameterError                            Traceback (most recent call last)
<ipython-input-7-3776517cab8b> in <module>()
      1 onechan_signal = np.random.randn(1, 1000)
----> 2 resampled_onechan = librosa.resample(onechan_signal, 1000, 2000)
      3 print(resampled_onechan.shape)

1 frames
/usr/local/lib/python3.7/dist-packages/librosa/util/utils.py in valid_audio(y, mono)
    304     elif y.ndim == 2 and y.shape[0] < 2:
    305         raise ParameterError(
--> 306             "Mono data must have shape (samples,). " "Received shape={}".format(y.shape)
    307         )
    308 

ParameterError: Mono data must have shape (samples,). Received shape=(1, 1000)
```

**Software versions**
```
Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
Python 3.7.10 (default, May  3 2021, 02:48:31) 
[GCC 7.5.0]
NumPy 1.19.5
SciPy 1.4.1
librosa 0.8.1
INSTALLED VERSIONS
------------------
python: 3.7.10 (default, May  3 2021, 02:48:31) 
[GCC 7.5.0]

librosa: 0.8.1

audioread: 2.1.9
numpy: 1.19.5
scipy: 1.4.1
sklearn: 0.22.2.post1
joblib: 1.0.1
decorator: 4.4.2
soundfile: 0.10.3
resampy: 0.2.2
numba: 0.51.2

numpydoc: None
sphinx: 1.8.5
sphinx_rtd_theme: None
sphinxcontrib.versioning: None
sphinx-gallery: None
pytest: 3.6.4
pytest-mpl: None
pytest-cov: None
matplotlib: 3.2.2
presets: None
```

**Comment by bmcfee** [comment]
This ties into a broader issue around valid audio that we're currently reasoning out in #1351 (implementing #1130 ): what shapes are accepted for "valid audio"?  Are any shapes valid?

Originally, valid_audio was meant to check several things in one go:
- contiguity in memory
- finite values
- floating point dtype
- appropriate shape `(n_samples,)` (mono) or `(2, n_samples)` (stereo).

As the library has developed, some of those restrictions have been relaxes.  For example, we no longer require memory contiguity, though we exploit it if it's there (when framing).

Finitude and floating point should probably stay, but I'd be curious about arguments to the contrary.  Do we want to propagate nans?  Do we want to accept complex signals?  (If the latter, do we need to do full spectrum STFTs when inputs are complex?)

Shape is a strange one.  We originally designed only for mono signals.  We then extended limited support for stereo.  #1351 is generalizing to fully multi-channel, and in doing so, I think we should also accept `(1, n_samples)` as a valid shape.  In the comments thread of #1351 , I also mention use cases for more than two dimensions, eg, `(n_sensors, n_channels, n_samples)`.  It's a bit more work, but not substantially more, to support that above stereo signals.  So the question is: do we need shape checks at all?  Should "valid audio" just mean: finite and floating point?

**Comment by dpwe** [comment]
We need to be able to determine which dimension is time.  And since there is not universal agreement around which dimension that should be in, for instance, a two-dimensional signal (ahem), it's valuable to be able to flag and/or reject signals that appear to be transposed (i.e., the non-librosa format of (num_samples, num_chans)).  So that is the case for a "shape check", although it's a bit fraught to decide where to set the limit on num_chans (4? 16? 256? < num_samples?)

But in ML workflows, it's quite common to have things like (num_batch, num_channels, num_samples), so I think being able to work over arbitrary dimensionalities is useful.  In fact, stipulating that the last dimension is always considered time, and that you can do what you want ahead of that, finally provides a justification for the (num_channels, num_samples) librosa convention.

**Comment by bmcfee** [comment]
>  So that is the case for a "shape check", although it's a bit fraught to decide where to set the limit on num_chans (4? 16? 256? < num_samples?)

Indeed.  I guess I would expect the time-like dimension to have the largest size.  Perhaps this should relax to a warning if that check fails?



> But in ML workflows, it's quite common to have things like (num_batch, num_channels, num_samples), so I think being able to work over arbitrary dimensionalities is useful.

Yes, this is also a case that I have in mind.  It's a little bit of a stretch, as embedding librosa functions into something like a training loop / batch generator isn't exactly a natural thing to do, but we may as well support it.

I guess that settles it for me:

1. Valid audio should check finitude and dtype
2. Shape can be anything
3. If the trailing dimension is not the largest, throw a warning

Sound good?  Curious for input from @lostanlen and @iranroman

**Comment by lostanlen** [comment]
> Valid audio should check finitude and dtype
> Shape can be anything
> If the trailing dimension is not the largest, throw a warning


sounds good to me

**Comment by bmcfee** [comment]
I'll roll this in as part of #1351 .  I started to implement it as a separate branch, but realized that it will break things until we finish adding multichannel support everywhere else.

Because not every function will be able to support multichannel, I think it's still handy to keep `mono=` around as an option, but we'll need to be a bit careful with the semantics.  `(1, N)` is, I think, a valid mono shape (as is `(1,1,1,1,N)` or anything else where `y.size = y.shape[-1]`), but having an `ndim=1` check might still be handy for certain functions that really do require it.  One compromise solution would be to extend `valid_audio` to have a `strict=` flag with the following logic:

```python
if mono:
   if strict and y.ndim != 1:
       # Error: too many dimensions
   if not strict and y.shape[-1] != y.size:
       # Error: more than one non-singleton dimension
```

**Comment by lostanlen** [comment]
@bmcfee sorry if i'm late to this debate but what happens if the shape is `(K, n_samples)` with `K=0`? Or more generally `x.shape[i]==0` for some `i < len(x.shape)` ?

In NumPy the FFT behaves just as we'd expect:
```
>>> np.fft.fft(np.random.randn(1, 2, 3, 0, 256), n=512)
array([], shape=(1, 2, 3, 0, 512), dtype=complex128)
```

and likewise for matrix multiplication:
```
>>> x = np.random.randn(0, 3)
>>> np.dot(x.T, x)
array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])
```

So i see no reason why built-in librosa pipelines like MFCC couldn't support multidimensional empty arrays, while preserving the `shape` pattern and `dtype` (which is convenient when processing many waveforms and not worrying that some of them could be empty). By that logic, `x = np.random.randn(1, 2, 3, 0, 256)` should pass `valid_audio` (it currently doesn't).

Or should we disallow empty arrays? I would be fine with that, but that means that @bmcfee's `shape can be anything` motto needs to be taken with a grain of salt (and documented accordingly)

**Comment by bmcfee** [comment]
> So i see no reason why built-in librosa pipelines like MFCC couldn't support multidimensional empty arrays, while preserving the `shape` pattern and `dtype` (which is convenient for processing many waveforms in parallel and not worrying that some of them might be empty). By that logic, `x = np.random.randn(1, 2, 3, 0, 256)` should pass `valid_audio` (it currently doesn't).

For inputs that are expected to be audio, this should be fine in most places.  Some exceptions might be lpc, preemphasis, iirt, etc; because it will matter how exactly we handle the base case for recurrent processing.

For other types of data (eg spectrograms), it will matter where the 0 occurs because we sometimes need to interpret the shape of different axes.

I think the more worrisome case is where you have an array of shape `(num_channels, num_samples=0)`.  If this passes valid_audio, and goes into frame-based processing with `center=True`, then the result will be non-empty due to padding.  This isn't necessarily wrong, but it might be unexpected behavior that's difficult to explain.

**Comment by lostanlen** [comment]
Interesting. I wasn't suggestint that we'd validate (via `valid_audio`) empty arrays with no time samples, like `(1, 2, 3, 4, 0)`. Only those with nonzero timesamples and zero channels (whatever non-last dimension you assign to channel): like, `(1, 2, 3, 0, 256)`.

I agree with your remarks about recursive or autoregressive functions, like LPC, preemphasis, iirt, PCEN, etc. I also agree on the padding problem, which would work in zero mode but not in reflect mode. NumPy already covers this limitation nicely:

```
x = np.array([])
y = np.pad(x, 1, mode="constant")   # no error. y = array([0., 0.])
y = np.pad(x, 1, mode="reflect")       # ValueError: can't extend empty axis 0 using modes other than 'constant' or 'empty'
ValueError: can't extend empty axis 0 using modes other than 'constant' or 'empty'
```

So in my opinion, `np.array([])` should not pass `valid_audio` because we can't guarantee that a downstream STFT with `pad_mode=reflect` (the default!) will also pass NumPy's `np.pad`.

Back to my original question, i think it would be nice if the following assertions could hold:

```
x = librosa.mfcc(np.zeros((2, 1024)))
y = librosa.mfcc(np.zeros((0, 1024)))

assert x.dtype == y.dtype
assert x.shape[1:] == y.shape[1:]
```

**Comment by bmcfee** [comment]
> Interesting. I wasn't suggesting that we'd validate (via `valid_audio`) empty arrays with no time samples, like `(1, 2, 3, 4, 0)`. Only those with nonzero time samples and zero channels (whatever non-last dimension you assign to channel): like, `(1, 2, 3, 0, 256)`.

So an empty dimension is okay as long as it's not in the time position?  I guess the shape check would then be `y.shape[-1] != 0`?  



> I also agree on the padding problem, which would work in zero mode but not in reflect mode.
> So in my opinion, `np.array([])` should not pass `valid_audio` because we can't guarantee that a downstream STFT with `pad_mode=reflect` (the default!) will also pass NumPy's `np.pad`.

Or, we could switch the default padding behavior to constant/zeros.  It's going into a major revision, after all. :grin: 



> Back to my original question, i think it would be nice if the following assertions could hold:
> 
> ```python
> x = librosa.mfcc(np.zeros((2, 1024)))
> y = librosa.mfcc(np.zeros((0, 1024)))
> 
> assert x.dtype == y.dtype
> assert x.shape[1:] == y.shape[1:]
> ```

Yeah, I think I agree with that.  The lingering question for me is whether there are legitimate use cases for a length-0 signal.  I can't think of one right now, but there might be some weird edge cases that pop up, eg, in the cqt module where it might be nice to have a graceful fallback when the signal is sliced too thin.

**Comment by bmcfee** [comment]
Coming back on this one now that #1351 is near completion.  At this point in the multichannel branch, we never directly call `valid_audio` with `mono=True` (implicitly or explicitly).  All functions that accept time-domain signal inputs work with multichannel, so there's never a need to restrict to mono.  For that reason, I think we can deprecate the `mono=` parameter from `valid_audio` (which I'll do in a separate PR).

One slight caveat here is `waveshow`, which behaves slightly differently depending on the number of channels being 1 or 2, but that's not a question of validity.

I think we've resolved the original issue that triggered this thread, but there are still some remaining comments to wrap up.

On the question of length-0 axes, I think these should not count as "mono":
![image](https://user-images.githubusercontent.com/1190540/132051585-4ed5b4bb-d532-4efd-9d60-9dc9109ca5e4.png)

This clarifies the logic of what "mono" really means then, and it can be detected by:
```python
# All leading dimensions have length 1
mono = np.allclose(y.shape[:-1], 1)
```

Then there's the question of an array of shape `(..., 0, ...)` being "valid audio".  If we agree to the premise that the other remaining shape checks no longer apply:
https://github.com/librosa/librosa/blob/9732ae0775d359e24172bd396e98b44b7cddf19a/librosa/util/utils.py#L292-L307
then the only remaining test outside of types is finitude.  This is a "for all" kind of statement, so it should only fail in the presence of a non-finite counter-example.  An empty array has no such samples, so I think we should count empty arrays as valid for the sake of logical consistency.

As @lostanlen mentions, this could have some unintended consequences for downstream processing, eg, `stft` with reflection padding will fail when the input is empty.  I'm not aware of other such cases, but it wouldn't surprise me if they exist.  In my mind, the easiest solution here is to change the default padding to `mode='constant'` (zero padding), as proposed in my last comment.  We could (should) also expand the test coverage to hit every sensible function with a length-0 input signal just to see what happens.

Does the above sound reasonable?

**Comment by lostanlen** [comment]
Yes.

**Comment by bmcfee** [comment]
Ok, so three to-dos from this thread:

1. [ ] remove shape checks from `valid_audio` (maybe keep one for ndim>0).  I'll do this in #1351 
2. [ ] deprecate `mono=` parameter in 0.9, remove by next major (0.10 or 1.0). This will be a separate PR.
3. [ ] move stft edge padding to constant instead of reflect for the 0.9 release. This will be a separate PR.
